### PR TITLE
Use and semantics when searching for text

### PR DESCRIPTION
**Pull request by adfoster-r7**
When searching for multiple text terms, the search will now require for all words to be matched

### Before

When searching for `search postgresql login` - all modules matching either `postgresql` or `login` will be returned:
 
```
msf6 exploit(windows/smb/psexec) >  search postgresql login

Matching Modules
================

   #    Name                                           Disclosure Date  Rank       Check  Description
   -    ----                                           ---------------  ----       -----  -----------
   0    auxiliary/admin/appletv/appletv_display_image                   normal     No     Apple TV Image Remote Control
   1    auxiliary/admin/appletv/appletv_display_video                   normal     No     Apple TV Video Remote Control
 
    ...  many modules later ...

   248  post/windows/manage/sticky_keys                                 normal     No     Sticky Keys Persistance Module
   249  post/windows/manage/wdigest_caching                             normal     No     Windows Post Manage WDigest Credential Caching


Interact with a module by name or index, for example use 249 or use post/windows/manage/wdigest_caching
```

### After

When searching for `search postgresql login` - only modules matching both `postgresql` and `login` will be returned:

```
msf6 exploit(windows/tftp/distinct_tftp_traversal) > search postgresql login
Matching Modules
================
   #  Name                                       Disclosure Date  Rank    Check  Description
   -  ----                                       ---------------  ----    -----  -----------
   0  auxiliary/scanner/postgres/postgres_login                   normal  No     PostgreSQL Login Utility
```

### 1 - Console search

Open the console, and ensure that normal search functionality works as expected:

```
search postgres
search login
search postgres login
search author:egypt arch:x64
search psexec author:egypt arch:x64
```

### 2 - rpcd

Create an instance of msfrpcd in one tab:

```
bundle exec ruby ./msfrpcd -P foo -f
```

In another create the client:
```
bundle exec ./ruby msfrpc -P foo -a 127.0.0.1
```

Ensure search works as expected, and aligns with the existing msfconsole search functionality:
```
rpc.call("module.search", "postgres")
rpc.call("module.search", "login")
rpc.call("module.search", "postgres login")
rpc.call("module.search", "author:egypt arch:x64")
rpc.call("module.search", "psexec author:egypt arch:x64")
```

### 3 - json web service

The existing webservice should not be impacted by this code change, but for a sanity check - run the service with:

```
bundle exec thin --rackup msf-json-rpc.ru --address localhost --port 8081 --environment production --tag msf-json-rpc start
```

Run some search queries:

```
curl --request POST --url http://localhost:8081/api/v1/json-rpc --header 'content-type: application/json' --data '{ "jsonrpc": "2.0", "method": "module.search", "id": 1, "params": ["postgres login"] }'
curl --request POST --url http://localhost:8081/api/v1/json-rpc --header 'content-type: application/json' --data '{ "jsonrpc": "2.0", "method": "module.search", "id": 1, "params": ["author:egypt arch:x64"] }'
curl --request POST --url http://localhost:8081/api/v1/json-rpc --header 'content-type: application/json' --data '{ "jsonrpc": "2.0", "method": "module.search", "id": 1, "params": ["psexec author:egypt arch:x64"] }'

```

### CODE REVIEW

File: spec/lib/msf/core/modules/metadata/search_spec.rb [hunk 1]
```diff
@@ -203,6 +205,49 @@ def inverse_query_terms(search_string)
...
+            "author:Istv\xE1n",
+          ],
+          :reject => [
+            "author:István",
+            "author:Istv\u00E1n ",
+          ],
+          :test_inverse => false
+        )
+      end
+    end
```
Comment (adfoster-r7) [inline spec/lib/msf/core/modules/metadata/search_spec.rb:249]: Adding some tests around https://github.com/rapid7/metasploit-framework/pull/13282

File: lib/msf/core/modules/metadata/search.rb [hunk 2]
```diff
@@ -192,19 +211,30 @@ def is_match(params, module_metadata)
...
   end
 
+  def as_regex(search_term)
+    # Convert into a case-insensitive regex
+    utf8_buf = search_term.dup.force_encoding('UTF-8')
+    if utf8_buf.valid_encoding?
+       Regexp.new(Regexp.escape(utf8_buf), Regexp::IGNORECASE)
+    else
+      # If the encoding is invalid, default to a regex that matches anything
+      //
```
Comment (adfoster-r7) [inline lib/msf/core/modules/metadata/search.rb:234]: This defaulting behaior was requested to be implemented as part of https://github.com/rapid7/metasploit-framework/pull/13282 such that if there's an encoding issue - it's silently hidden from the user. Following the zen of python I'm tempted to raise an exception instead

Comment (dwelch-r7) [inline lib/msf/core/modules/metadata/search.rb:234]: Should this be a `Regex` object aswell like we have above?

**Review by dwelch-r7 — APPROVED** [review]
One minor question but LGTM, I'm happy enough for the search semantics to change, it makes more sense in my opinion

File: lib/msf/core/modules/metadata/search.rb (hunk 2 above)
Comment (adfoster-r7) [reply lib/msf/core/modules/metadata/search.rb:234]: @dwelch-r7 Regex is a first class citizen in Ruby's syntax, so `/.../` is actually just syntactic sugar for Regex.new.
However you can use `Regexp.new` if you want to convert a string to a regex.

So in this case, an empty regex of `//` is perfectly cromulent

**Comment by wvu-r7** [comment]
This looks like a regression in _behavior_. See #7647.

**Review by wvu — APPROVED** [review]
Didn't look too closely at the code, but the expected behavior is correct.

File: lib/msf/core/modules/metadata/search.rb (hunk 2 above)
Comment (dwelch-r7) [reply lib/msf/core/modules/metadata/search.rb:234]: ah I didn't know that, neat! I'm for sure glad the comment is there though, if nothing else but to remind me that # is a comment in ruby not // 😄

**Comment by dwelch-r7** [comment]
Nice, just gave it a quick test in the console to double check the highlighting still works, and it does!

**Comment by busterb** [comment]
Alan and I looked at this a couple of months ago, and I don't think current behavior of the module cache is a regression from the old DB based search either. It didn't appear to get fixed in #7647 for free-text search in the first place, only working if you used specific field-based terms like those mentioned in the old PR.

Thanks @afoster-r7, much needed improvement!

**Comment by adfoster-r7** [comment]
When dogfooding this I spotted a small last minute bug with the fall through search functionality of the `use` command - thanks for taking a look and fixing the issue (https://github.com/adfoster-r7/metasploit-framework/pull/2) @dwelch-r7 !

Landing this now 🚢

**Comment by adfoster-r7** [comment]
## Release Notes

Updates the search functionality for modules to require all text search terms to be matched. Now when searching for `search postgresql login` - only modules matching both `postgresql` and `login` will be returned.

**Pull request merged**